[HTTP 이해 - 웹 서버 구현] 2단계 - HTTP 웹 서버 구현

Wu22e opened the pull request:
안녕하세요 리뷰어님!

2단계 미션 진행하면서 요구사항 1 ~ 7 에 대한 내용을 학습하고 구현 하면서 
RequestHandler 객체의 복잡도가 너무 커져서 이 부분을 리팩토링하다 보니 미션 리뷰요청이 늦어 졌네요.. 

---

리팩토링하면서 궁금한 점이 하나 있었는데,

HttpResponse 객체를 생성 할 때, 응답 프로토콜, 응답 상태코드, 헤더 셋팅, 바디 값을 파라미터로 받아 생성하다 보니 
클라이언트에서 HttpResponse를 생성할 때 복잡도가 증가하였습니다.

그래서 현재 각 Controller 구현체들의 복잡도나 중복이 좀 많다고 느껴졌는데, 이런 부분에 대해서는 HttpResponse 객체가 담당할 수 있도록 책임을 좀더 가져가는게 맞을지 리뷰어님의 생각이 궁금합니다.

리뷰 요청이 많이 늦어졌지만.. 앞으로 힘내서 열심히 해보도록 하겠습니다.
리뷰 잘부탁드립니다!

Wu22e committed:
refactor : Path QueryString nullable -> 빈 값이 들어가도록 변경

Wu22e committed:
refactor : 문자열 상수화

Wu22e committed:
docs : 기능 요구사항 정리

Wu22e committed:
feat : 요구사항 1 - index.html 파일 응답

Wu22e committed:
refactor : QueryString non-null 필드로 인한 equals & hashCode 재정의

- setQueryString 메서드 value 없을 경우 로직 처리 추가

Wu22e committed:
test : file path 존재 여부 테스트

Wu22e committed:
feat : 요구사항 2 - 회원가입(GET)

Wu22e committed:
feat : 요구사항 3 - 회원가입(POST)

- 패키지 추가 및 파일 이동
- Method equal 메서드 추가

Wu22e committed:
feat : 요구사항 4 - 회원가입 후 redirect

Wu22e committed:
docs : 기능 요구사항 정리

Wu22e committed:
feat : 요구사항 5 - 로그인 결과에 따른 처리

- HttpRequest 객체 & 테스트 추가

Wu22e committed:
feat : 요구사항 6 - 로그인 상태에 따른 /user/list 접근 권한 처리

- Cookie 객체 생성
- HandleBars 사용

Wu22e committed:
feat : 요구사항 7 -  Stylesheet 파일 지원

Wu22e committed:
docs : 기능 요구사항 추가 사항 정리

Wu22e committed:
refactor : HttpResponse 객체 생성 & HttpRequest 리팩토링

- 패키지 정리
- 응답 처리를 위한 getter 추가
- 네이밍 변경 / 생성자 추가 / match 메서드 추가
- 응답 처리를 위한 StatusCode / StatusLine 객체 생성

Wu22e committed:
refactor : 요청 분기 처리 -> 컨트롤러 요청 매핑 로직 추가

- Controller 인터페이스 추가
- 각 요청별 Controller 구현체 추가

Wu22e committed:
refactor : 요청 분기 맵핑 로직 추가 & RequestHandler 리팩토링

- RequestMapping 객체 생성

Wu22e committed:
refactor : ProtocolType 객체 추가

Wu22e committed:
refactor : header value 존재하지 않는 key 에 대한 처리 추가

kang-hyungu commented on src/main/java/webserver/RequestMapping.java:
```diff
@@ -0,0 +1,55 @@
+import java.util.Set;
+
+public class RequestMapping {
+    private static Set<Controller> CONTROLLERS = new HashSet<>();
```
final을 붙여도 괜찮겠네요

```suggestion
    private static final Set<Controller> CONTROLLERS = new HashSet<>();
```

kang-hyungu commented on src/main/java/webserver/controller/StaticController.java:
```diff
@@ -0,0 +1,28 @@
+    @Override
+    public HttpResponse process(HttpRequest httpRequest) throws IOException, URISyntaxException {
+        return HttpResponse.ok(
+                new Header(Map.of("Content-Type", "text/css;charset=utf-8")),
```
http header에서 자주 쓰이는 텍스트는 enum으로 관리하면 중복 코드를 줄일 수 있겠네요

kang-hyungu commented on src/main/java/webserver/controller/UserListController.java:
```diff
@@ -0,0 +1,50 @@
+import java.util.Map;
+
+public class UserListController implements Controller {
+    private final static String TEMPLATE_ENGINE_USERS_KEY = "users";
```
```suggestion
    private static final String TEMPLATE_ENGINE_USERS_KEY = "users";
```

kang-hyungu commented on src/main/java/webserver/controller/UserListController.java:
```diff
@@ -0,0 +1,50 @@
+                                    "Set-Cookie", "logined=true; Path=/"
+                            )
+                    ),
+                    handlebars.compile(httpRequest.getPath()).apply(Map.of(TEMPLATE_ENGINE_USERS_KEY, DataBase.findAll())).getBytes());
```
한 줄로 모든 처리를 하게 되면 협업 할 때 코드 파악이 어려워집니다.
적절한 변수를 추출해서 어떤 작업인지 파악하기 용이하게 만들어보시기 바랍니다.

kang-hyungu commented on src/main/java/webserver/controller/UserLoginController.java:
```diff
@@ -0,0 +1,47 @@
+import java.util.Map;
+
+public class UserLoginController implements Controller {
+    private static final String TEMPLATE_PATH = "./templates";
```
불필요한 변수는 제거하는게 좋겠네요

kang-hyungu commented on src/main/java/webserver/controller/UserLoginController.java:
```diff
@@ -0,0 +1,47 @@
+                    new Header(
+                            Map.of(
+                                    "Content-Type", "text/html;charset=utf-8",
+                                    "Set-Cookie", "logined=false; Path=/"
```
컨트롤러마다 반복해서 헤더 처리를 위한 문자열이 반복해서 나오고 있네요
중복을 제거할 방법을 찾아보아요.

kang-hyungu commented on src/main/java/webserver/http/Header.java:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    public Header add(String key, String value) {
+        Map<String, String> fields = new HashMap<>(this.fields);
```
this.fields를 map으로 다시 감쌀 필요가 있을까요?

Wu22e replied:
해당 부분의 경우, 클라이언트에서 Header 객체를 생성할때 Map.of() -> immutable map을 사용하다 보니 새로운 값들이 추가되지 않아 새로운 HashMap으로 mutable 하게 감쌌습니다. 혹시 더 나은 방법이 있을까요?

kang-hyungu replied:
불변맵을 사용해야될 이유가 없다면 mutable map을 사용하면 되지 않을까요?

kang-hyungu commented on src/main/java/webserver/http/request/HttpRequest.java:
```diff
@@ -0,0 +1,142 @@
+        }
+    }
+
+    public static HttpRequest of(BufferedReader br) throws IOException {
```
정적 팩터리 메서드보다 생성자를 활용해도 되지 않을까요?

Wu22e replied:
정적 팩터리 메서드를 사용하는 이유가 객체 생성에 대한 이름을 보기좋게 부여하거나 생성자에서는 값 셋팅만하고 정적 팩터리 메서드 에서는 validation 체크, 혹은 객체 에 값을 셋팅하기 위해 값을 정제하는 과정으로 알고있는데, 혹시 제가 잘못 알고 있는 부분이 있을까요?

저는 값을 정제하여 객체 생성시 값을 정해 준다는 생각을 가지고 정적 팩터리 메서드를 만들게 되었습니다. 
제가 피드백의 의도를 잘 파악하지 못하여 질문 드립니다!

kang-hyungu replied:
생성자에서 값 셋팅만 해야되는 이유가 무엇인가요?
클라이언트 입장에서 생성자는 값 셋팅 용도로만 한정해서 사용한다는 내용을 알고 있어야 하는 문제가 있겠네요.
말씀하신 유효성 체크, 값 정제는 생성자에서 처리해도 문제는 없습니다. 생성자만으로 어떤 객체를 생성하는지 파악하기 어려울 때 정적 팩터리 메서드를 활용하시는 걸 추천합니다.
관련 내용은 1주차 자바지기님 강의 때도 언급되었으니 참고하세요.

kang-hyungu commented on src/main/java/webserver/http/response/HttpResponse.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    public static HttpResponse ok(Header header, byte[] body) {
+        return new HttpResponse(StatusLine.of(new Protocol(ProtocolType.HTTP, Version.ONE_ONE), StatusCode.OK), header, body);
```
한 줄에서 너무 많은 처리를 하고 있네요.
정적 팩터리 메서드를 차라리 이렇게 써보면 어떨까요?

```suggestion
        return new HttpResponse(StatusLine.ofHttp11(), header, body);
```

kang-hyungu reviewed (changes requested):
안녕하세요 형우님
몇 가지 챙겨보면 좋을 부분에 코멘트 남겼습니다.
수정해보시고 리뷰 재요청주세요.

Wu22e committed:
refactor : 불 필요한 상수 제거

Wu22e committed:
refactor : 헤더 키 & 값 -> Enum 및 상수로 관리

- Header 로그인 성공/실패, 정적파일/템플릿파일 요청에 따른 메서드 추가 (중복 제거)
- Header 생성 null validate 추가

Wu22e committed:
refactor : Util RequestMapping 클래스 private 생성자 추가

Wu22e committed:
refactor : validate 추가 & 정적 팰터리 메서드를 이용하여 클라이언트 코드 중복 제거

Wu22e committed:
refactor : 변수 및 메서드를 추출 (읽기 쉽도록 리팩토링)

Wu22e commented:
안녕하세요 리뷰어님!
피드백 반영하여 리뷰 재요청 드립니다!
피드백에 대하여 궁금한 내용들을 댓글로 작성하였습니다. 

부족한 부분이 있다면 피드백 다시 주시면 감사하겠습니다!
리뷰 잘부탁드립니다!

kang-hyungu commented on src/main/java/webserver/http/HeaderKey.java:
```diff
@@ -0,0 +1,47 @@
+
+import java.util.Arrays;
+
+public enum HeaderKey {
```
모든 http 헤더를 한 enum으로 관리하니 복잡하네요.
generic, request, response, entity 헤더로 나눌 수 있으니 참고하세요.
https://developer.mozilla.org/ko/docs/Web/HTTP/Headers

kang-hyungu reviewed (approved):
형우님 피드백 반영 잘 해주셨습니다.
pr은 머지 처리할게요 다음 미션 진행해주세요.
